src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -298,120 +298,110 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((surrogateCalculation >> 12) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((surrogateCalculation >> 6) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((surrogateCalculation) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
+                                goto need_more; // Invalid surrogate pair.
                             }
-                            else if (codePoint <= Utf8ThreeBytesLastCodePoint)
+
+                            int surrogateCalculation;
+                            unchecked
                             {
-                                if (bytesWrittenTemp > outputBytesAvailable - 3)
-                                {
-                                    goto need_more;
-                                }
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint >> 12) & b0000_1111U) | b1110_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint >> 6) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
+                                surrogateCalculation = ((lowSurrogate - Utf16LowSurrogateFirstCodePoint)
+                                    | ((codePoint - Utf16HighSurrogateFirstCodePoint) << 10)) + 0x10000;
                             }
-                            else
+
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation >> 18) & b0000_0111U) | b1111_0000U);
+                            bytesWrittenTemp++;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation >> 12) & b0011_1111U) | b1000_0000U);
+                            bytesWrittenTemp++;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation >> 6) & b0011_1111U) | b1000_0000U);
+                            bytesWrittenTemp++;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((surrogateCalculation) & b0011_1111U) | b1000_0000U);
+                            bytesWrittenTemp++;
+                        }
+                        else
+                        {
+                            if (bytesWrittenTemp > outputBytesAvailable - 3)
                             {
-                                throw new ArgumentException("Invalid surrogate pair.", nameof(utf16));
+                                goto need_more;
```

Comment:
For this specific case, I realized we will never reach this else branch (hence I removed it). We already have a multiplexer for all char values between 0-FFFF in other if/else branches. 

Generally, talking with Robert/Krzysztof, we just return false and how much was consumed rather than throwing, when we get invalid surrogate pairs. The fallback/recovery would be built on top.